### DISCUSSION TITLE
Configure Xrdp to serve different Apps depending on the menu (Session) chosen in the greeter

### DISCUSSION THREAD

**Discussion by vongillus** [neutrinolabs/xrdp#2527] [Q&A] [answered]
Hi,

I have tried the whole afternoon to find a way, but I am out of ideas now. I serve through Xrdp/Xvnc an application which is started through a script. That script is configured in sesman.ini as the "DefaultWindowManager". This works great.

What I would like to achieve is depending on the menu chosen in the Xrdp Greeter (called Session), start a different application. So if I choose the first menu item, it would call one "DefaultWindowManager" and if I choose the second menu item, it would call another "DefaultWindowManager".
Another way would be to pass a variable to anything before the Window Manager script is called, so that I can handle that from within my script:

![image](https://user-images.githubusercontent.com/80924891/215821533-75a59dd6-3709-4499-b7e2-e23c46867469.png)

Any way to do this (I am running RHEL8)?

Thanks

**Comment by akarl10** [comment, marked as answer]
as far as I know you there is no information around what the user selected when startwm is called.
I agree that it would be cool if the desktop environment could be selected here.

I think you knew that already, but rdp has a "shell" parameter

with mstsc you can pass that parameter by creating a .rdp file and appending ex.
```
alternate shell:s:/etc/xrdp/my-custom-shell.sh
```
with freerdp the parameter is `/shell:/etc/xrdp/my-custom-shell.sh`

then instead of startwm.sh this command is executed

**Comment by vongillus** [comment]
thanks for your suggestion, it would be an option even if I would have to deploy then that rdp file to clients. Better than nothing!

I tried the option "[SessionVariables]" which I can see from from my script but it doesn't tell me which Session was selected. I will have to look in the code may be.

**Comment by matt335672** [comment]
It's pretty straightforward to pass the session type through to the session startup script as an environment variable, and it would certainly help this use-case. Are you in a position to build from source, @vongillus ?

**Comment by vongillus** [comment]
Well I can for sure build from source since I would deploy this specific XRDP version only on a bunch of servers, but for this I would need a patch. And that's where I am completely out of my field. So if you have any idea, I would really appreciate! :)

**Comment by matt335672** [comment]
Thanks. What version of xrdp are you planning on deploying? We can patch against that.

**Comment by vongillus** [comment]
I am planning to recompile the one which I use from EPEL, which is https://download.fedoraproject.org/pub/epel/8/Everything/SRPMS/x/xrdp-0.9.21-1.el8.src.rpm
So probably it's version 0.9.21 if nothing was backported to it. I would download the source RPM, patch it and recompile it. Does this sounds good?

**Comment by matt335672** [comment]
Hi @vongillus 

I've just had a more detailed look at this, and I must apologise for leading you on. It's way more complicated to do this for v0.9.x because the middleware for that version is rather difficult to modify without lots of testing. We replaced it completely in the devel version, and that's what I was thinking of when I replied to you (rather too hastily it seems).

The implementation of this feature needs the session type string to be passed from xrdp to xrdp-sesman.

I'll certainly mark this as a feature request for the next major release. In the meantime, I think @akarl10's suggestion is probably the way to go, as a patch from me would be rather complex and not tested to the same degree as the vanilla code from EPEL.

**Reply by vongillus** [reply]
Hi @matt335672 

no problem, it's not like I order and you have to implement! Yes, I will follow suggestion from @akarl10 then. Thanks anyway for being so reactive, I was amazed. One can do really great things with XRDP, it's incredible.
